### PR TITLE
[ISSUE-269] Use CR caching in csi-node Discover loop

### PR DESCRIPTION
**Pull request by ykulazhenkov**
### Issue https://github.com/dell/csi-baremetal/issues/269

Integrate cache CRHelper to volumemgr

All CRs read requests from csi-node Discover loop show receive cached data from the auto-updated cache

## PR checklist
- [x] Add link to the issue
- [x] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

### CODE REVIEW

File: pkg/node/node.go [hunk 1]
```diff
@@ -26,6 +26,7 @@ import (
 	"sync"
 
 	"github.com/container-storage-interface/spec/lib/go/csi"
+	csibmnodeconst "github.com/dell/csi-baremetal/pkg/crcontrollers/csibmnode/common"
```
Comment (AnastasiaShemyakinskaya) [inline pkg/node/node.go:29]: Wrong import format

File: cmd/node/main.go [hunk 2]
```diff
@@ -115,8 +117,22 @@ func main() {
...
+	wrappedK8SClient := k8s.NewKubeClient(k8SClient, logger, *namespace)
+
+	k8sCache, err := k8s.GetK8SCache()
+	if err != nil {
+		logger.Fatalf("fail to create cache for kubernetes resources, error: %v", err)
+	}
+	// start cache
+	go func() {
+		// cache implementation we use newer returns err
+		_ = k8sCache.Start(stopCH)
```
Comment (mishoyama) [inline cmd/node/main.go:129]: If caching go routine died will we start using non-cached mode, read outdated information or just fail?

File: pkg/base/k8s/kubecache.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	k8sCl "sigs.k8s.io/controller-runtime/pkg/client"
+	crApiutil "sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+// GetK8SCache returns k8s cache
+func GetK8SCache() (cache.Cache, error) {
+	config := ctrl.GetConfigOrDie()
+	scheme, err := PrepareScheme()
```
Comment (mishoyama) [inline pkg/base/k8s/kubecache.go:33]: Does it make sense to cache all CSI custom resources? Can we specify only particular?

File: pkg/base/k8s/kubecache.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// KubeCache is a wrapper for controller-runtime cache
+type KubeCache struct {
+	k8sCl.Reader
+	log *logrus.Entry
+}
+
+// ReadCR CRReader implementation
+func (k KubeCache) ReadCR(ctx context.Context, name string, namespace string, obj runtime.Object) error {
```
Comment (mishoyama) [inline pkg/base/k8s/kubecache.go:56]: How does caching work when we update CR though usual k8s client?

File: pkg/node/volumemgr.go [hunk 5]
```diff
@@ -699,7 +706,7 @@ func (m *VolumeManager) discoverVolumeCRs() error {
 		"method": "discoverVolumeCRs",
 	})
 
-	driveCRs, err := m.crHelper.GetDriveCRs(m.nodeID)
+	driveCRs, err := m.cachedCrHelper.GetDriveCRs(m.nodeID)
```
Comment (mishoyama) [inline pkg/node/volumemgr.go:709]: Disk can be marked as OFFLINE in controller **statemonitor.go**. Is this really safe to read from cache here?

File: cmd/node/main.go (hunk 2 above)
Comment (ykulazhenkov) [reply cmd/node/main.go:129]: Cache implements retry logic. If WATCH call to kube API fails, reconnection will happen automatically. If there is no connection at the moment old data still can be readable from the cache.

File: pkg/base/k8s/kubecache.go (hunk 3 above)
Comment (ykulazhenkov) [reply pkg/base/k8s/kubecache.go:33]: We don't cache all resources. It is just a schema preparation, so the client will know that these resources exist. Cache implementation will subscribe to resource updates only if this resource was read from the cache at least once.

File: pkg/node/volumemgr.go (hunk 5 above)
Comment (ykulazhenkov) [reply pkg/node/volumemgr.go:709]: Cache is autoupdated, so we will receive this update.

File: pkg/base/k8s/kubecache.go (hunk 4 above)
Comment (ykulazhenkov) [reply pkg/base/k8s/kubecache.go:56]: Cache updates happen automatically in background. This is fully separate process from updates from another client instances.

**Review by mishoyama — APPROVED** [review]

File: pkg/node/node.go (hunk 1 above)
Comment (ykulazhenkov) [reply pkg/node/node.go]: fixed

**Review by AnastasiaShemyakinskaya — APPROVED** [review]

**Review by painhardcore — APPROVED** [review]

**Pull request merged**